### PR TITLE
Add showLastCharWhenObscureText

### LINKED ISSUE #65352: Flutter app still shows leading password character on Android even after disabling feature in settings (Security Vulnerability)
## Steps to Reproduce

1. Run `flutter create bug`.
2. Update the files as follows:
```dart
// lib/main.dart
import 'package:flutter/material.dart';

void main() => runApp(App());

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: TextField(obscureText: true),
        ),
      ),
    );
  }
}

```
3. On Android, go to settings.
4. Go to 'Security' or 'Security & Location'
5. Uncheck 'Show Passwords' (Display characters briefly as you type)
6. Open Flutter app `bug` and type in any text

**Expected results:** 
There should be **no** characters displayed for any time period as password 

**Actual results:** 
Characters are **displayed** briefly as you type

<details>
  <summary>Logs</summary>

```
[√] Flutter (Channel stable, 1.20.3, on Microsoft Windows [Version 10.0.10240], locale en-US)
    • Flutter version 1.20.3 at C:\flutter
    • Framework revision 216dee60c0 (6 days ago), 2020-09-01 12:24:47 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\Users\Joseph\AppData\Local\Android\Sdk
    • Platform android-29, build-tools 29.0.2
    • ANDROID_SDK_ROOT = C:\Users\Joseph\AppData\Local\Android\Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[!] Android Studio (version 4.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code, 64-bit edition (version 1.48.2)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 3.14.0

[√] Connected device (1 available)
    • Infinix X610B (mobile) • 0425325972000854 • android-arm64 • Android 9 (API 28)

! Doctor found issues in 1 category.
```

</details>

### PR DESCRIPTION
**Pull request by zhangwei911**
*Add showLastCharWhenObscureText for TextField when obscureTextfield is set to true,it will show the last char,but in the bank app,this is not allow*

Fixes https://github.com/flutter/flutter/issues/65352

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by zhangwei911** [comment]
@googlebot I signed it!

**Review by justinmc — COMMENTED** [review]
Sorry for the slow review, this is an important bug to fix!

The main thing that this PR still needs is a test.  Could you add one to editable_text_test.dart?  There should be similar tests for obscureText that you can copy and/or modify.

Also, I want to start a discussion about the name showLastCharWhenObscureText and the use of two bools for it and obscureText.  What if instead we deprecated obscureText and created an enum with three values?  Would that make the situation any simpler?

File: packages/flutter/lib/src/material/text_field.dart [hunk 1]
```diff
@@ -321,7 +321,7 @@ class TextField extends StatefulWidget {
   /// to [ui.BoxHeightStyle.tight] and [ui.BoxWidthStyle.tight] respectively and
   /// must not be null.
   ///
-  /// The [textAlign], [autofocus], [obscureText], [readOnly], [autocorrect],
+  /// The [textAlign], [autofocus], [obscureText], [showLastCharWhenObscureText], [readOnly], [autocorrect],
```
[the remark this answers is not in the document]
Comment (zhangwei911) [reply packages/flutter/lib/src/material/text_field.dart:324]: i have update

**Comment by zhangwei911** [comment]
@justinmc like this?
```
enum Obscure {
    /// show all text
    none,
    /// obscure all text
    all,
   /// obscure all text but the last char
    default
}
```

**Comment by justinmc** [comment]
@zhangwei911 Yes that's what I was talking about.  I think that approach looks good.  Can we get rid of showLastCharWhenObscureText now, or is there a reason you kept it?

Also, I think we should change the name from `obscure` to `obscureTextBehavior` and the enum would be `ObscureTextBehavior`.  I think `none` and `all` are ok, but maybe `last` should change to be `delayed`?  Since really we are obscuring everything, just at a short delay.

CC @LongCatIsLooong for a second opinion on this API.

**Comment by justinmc** [comment]
Also FYI there is a conflict on this PR, so you'll have to merge/rebase master.

**Comment by zhangwei911** [comment]
@justinmc i have delete showLastCharWhenObscure,update enum and param name.
but flutter-build — Tree is currently broken. Please do not merge this PR unless it contains a fix for the tree.
i don`t know how to fix,its my first time to pull request

File: packages/flutter/lib/src/services/text_input.dart [hunk 2]
```diff
@@ -431,6 +431,15 @@ enum TextCapitalization {
   none,
 }
 
+enum ObscureTextBehavior {
+  /// show all text
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_input.dart:435]: We need full sentences for these, so maybe change to:

Indicates all text should be shown.

Indicates all text should be obscured.

Indicates all text should be obscured, but the last character is visible for a short delay.

File: packages/flutter/lib/src/services/text_input.dart [hunk 3]
```diff
@@ -431,6 +431,15 @@ enum TextCapitalization {
...
+  none,
+  /// obscure all text
+  all,
+  /// obscure all text but the last char just at a short delay
+  delayed
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_input.dart:440]: Comma here.

File: packages/flutter/lib/src/services/text_input.dart [hunk 4]
```diff
@@ -484,11 +495,24 @@ class TextInputConfiguration {
   /// Defaults to false.
   final bool readOnly;
 
+  @Deprecated(
+    'use obscure instead.'
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_input.dart:499]: Here and below, capitalize the "u", and change "obscure" to "obscureTextBehavior".

File: packages/flutter/lib/src/services/text_input.dart [hunk 5]
```diff
@@ -484,11 +495,24 @@ class TextInputConfiguration {
...
+  )
   /// Whether to hide the text being edited (e.g., for passwords).
   ///
   /// Defaults to false.
   final bool obscureText;
 
+  @Deprecated(
+    'use obscure instead.'
+  )
+
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_input.dart:509]: Remove this newline.

File: packages/flutter/lib/src/services/text_input.dart [hunk 6]
```diff
@@ -484,11 +495,24 @@ class TextInputConfiguration {
...
+  /// Whether or not the last character entered is shown before
+  /// being obscured.
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_input.dart:511]: This could probably use the flutter.widgets.editableText.obscureTextBehavior template.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 7]
```diff
@@ -572,6 +576,18 @@ class EditableText extends StatefulWidget {
...
+  /// Whether or not the last character entered is shown before
+  /// being obscured.
+  /// When this is set to [ObscureTextBehavior.all] or [ObscureTextBehavior.delayed], the last characters in the text field is
+  /// replaced by [obscuringCharacter].
+  ///
+  /// Defaults to [ObscureTextBehavior.none]. Cannot be null.
+  ///
+  /// {@endtemplate}
+  final ObscureTextBehavior obscureTextBehavior;
+
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:590]: Remove extra newline here.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 8]
```diff
@@ -572,6 +576,18 @@ class EditableText extends StatefulWidget {
...
   final bool obscureText;
+  
+  /// {@template flutter.widgets.editableText.obscureTextBehavior}
+  /// Whether or not the last character entered is shown before
+  /// being obscured.
+  /// When this is set to [ObscureTextBehavior.all] or [ObscureTextBehavior.delayed], the last characters in the text field is
+  /// replaced by [obscuringCharacter].
+  ///
+  /// Defaults to [ObscureTextBehavior.none]. Cannot be null.
+  ///
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:587]: Remove the newline here.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 9]
```diff
@@ -572,6 +576,18 @@ class EditableText extends StatefulWidget {
   /// Defaults to false. Cannot be null.
   /// {@endtemplate}
   final bool obscureText;
+  
+  /// {@template flutter.widgets.editableText.obscureTextBehavior}
+  /// Whether or not the last character entered is shown before
+  /// being obscured.
+  /// When this is set to [ObscureTextBehavior.all] or [ObscureTextBehavior.delayed], the last characters in the text field is
+  /// replaced by [obscuringCharacter].
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:584]: This is probably too specific, maybe change it to:

```
How characters in the field are obscured, if at all.

For example, a password field may want to obscure the entered text so it's not readable.

When this is set to [ObscureTextBehavior.all] or [ObscureTextBehavior.delayed], the characters in the field are replaced by [obscuringCharacter].
```

But make sure that lines don't exceed 80 characters in length.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 10]
```diff
@@ -499,6 +502,7 @@ class EditableText extends StatefulWidget {
        assert(focusNode != null),
        assert(obscuringCharacter != null && obscuringCharacter.length == 1),
        assert(obscureText != null),
+       assert(obscureTextBehavior != null),
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:505]: I think you can also add an assertion that both `obscureText` and `obscureTextBehavior` aren't used at the same time, to avoid confusion.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 11]
```diff
@@ -2846,6 +2846,61 @@ void main() {
     semantics.dispose();
   });
 
+  testWidgets('password fields hide the last char when obscure is ObscureTextBehavior.all', (WidgetTester tester) async {
```
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_test.dart:2849]: Can you also test ObscureTextBehavior.delayed?

**Review by justinmc — COMMENTED** [review]
Some cleanup to do here from my comments, but I like this approach now.  Thank you for the PR!  Let me know if my comments don't make sense.

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 17.
View them at https://flutter-gold.skia.org/cl/github/77518

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 21.
View them at https://flutter-gold.skia.org/cl/github/77518

File: packages/flutter/lib/src/material/text_field.dart [hunk 12]
```diff
@@ -351,6 +352,7 @@ class TextField extends StatefulWidget {
     this.autofocus = false,
     this.obscuringCharacter = '•',
     this.obscureText = false,
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_field.dart:354]: I think these other instances of `obscureText` should be @Deprecated as well, right?

Comment (justinmc) [reply packages/flutter/lib/src/material/text_field.dart:354]: Also in editable_text.dart, cupertino/text_field.dart, and anywhere else.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 13]
```diff
@@ -572,6 +575,18 @@ class EditableText extends StatefulWidget {
   /// Defaults to false. Cannot be null.
   /// {@endtemplate}
   final bool obscureText;
+  
+  /// {@template flutter.widgets.editableText.obscureTextBehavior}
+  /// How characters in the field are obscured, if at all.
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:580]: There should be an empty line in between paragraphs in the docs.

```dart
/// How characters in the field are obscured, if at all.
///
```

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 14]
```diff
@@ -2846,6 +2846,116 @@ void main() {
...
+        ),
+        ignoreTransform: true,
+        ignoreRect: true,
+        ignoreId: true,
+      ),
+    );
+
+    semantics.dispose();
+  });
+
```
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_test.dart:2958]: Could you also test that the text in the field is immediately changed to the obscuring character when set to `all`?  Something like this but without obscuringTextCharacter: https://github.com/flutter/flutter/blob/5acf7e98f16222610a4e6bfe1e36b08a90cf53a4/packages/flutter/test/widgets/editable_text_test.dart#L2961

**Review by justinmc — COMMENTED** [review]
Check the failing analyzer check, you've got a bunch of trailing spaces that need to be removed.

Also, the goldens are failing.  It looks like there is some layout overflow.  Maybe try running one of the failing golden tests locally and see if there is an error on the console?

**Comment by justinmc** [comment]
@zhangwei911 You've got analyzer errors because of your deprecation format, could you take a look? Thanks for updating the other stuff.  https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8841034535413644464/+/u/prepare_environment/Analyze/analyze/stdout

File: packages/flutter/lib/src/material/text_field.dart [hunk 15]
```diff
@@ -350,7 +351,12 @@ class TextField extends StatefulWidget {
     this.showCursor,
     this.autofocus = false,
     this.obscuringCharacter = '•',
+    @Deprecated(
+        'use obscureTextBehavior instead.'
+        'This feature was deprecated after v1.25.0-5.0.pre.'
```
[the remark this answers is not in the document]
Comment (zhangwei911) [reply packages/flutter/lib/src/material/text_field.dart]: it doesn`t work

[the remark this answers is not in the document]
Comment (zhangwei911) [reply packages/flutter/lib/src/material/text_field.dart]: it must add 1 space in the end of the first line and use the lastest dev version

**Comment by justinmc** [comment]
@zhangwei911 Can you fix the conflicts?  There are also lots of weird analyzer failures.  They'll hopefully go away when you fix the conflicts.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 16]
```diff
@@ -142,7 +142,10 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
     TextHeightBehavior? textHeightBehavior,
     TextWidthBasis textWidthBasis = TextWidthBasis.parent,
     String obscuringCharacter = '•',
-    bool obscureText = false,
+    @Deprecated(
+      'Use obscureTextBehavior instead. '
+      'This feature was deprecated after v2.6.0-0.0.pre.',
+    )
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/editable.dart:148]: I think you meant to keep `obscureText` and add the Deprecated flag above it?

**Review by justinmc — COMMENTED** [review]
@zhangwei911 Does this successfully run for you locally?  I think there are some errors that are tripping up the analyzer.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 16 above)
Comment (zhangwei911) [reply packages/flutter/lib/src/rendering/editable.dart:148]: yes,now it does not a break change.
i found many error files,but i does not edit it

**Comment by justinmc** [comment]
This still seems to have some problems, does it run locally for you?

**Comment by zhangwei911** [comment]
> This still seems to have some problems, does it run locally for you?

i have run locally
ObscureTextBehavior
ObscureTextBehavior1

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 17]
```diff
@@ -300,10 +305,12 @@ class CupertinoTextField extends StatefulWidget {
        assert(readOnly != null),
        assert(autofocus != null),
        assert(obscuringCharacter != null && obscuringCharacter.length == 1),
-       assert(obscureText != null),
+       assert(obscureText != null || obscureTextBehavior != null),
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_field.dart:308]: We should also make sure that people don't try to use both:

```dart
assert(obscureText == null || obscureTextBehavior == null);
```

Here and everywhere else that accepts obscureTextBehavior.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 18]
```diff
@@ -377,7 +393,8 @@ class CupertinoTextField extends StatefulWidget {
   /// must not be null.
   ///
   /// The [autocorrect], [autofocus], [clearButtonMode], [dragStartBehavior],
-  /// [expands], [maxLengthEnforced], [obscureText], [prefixMode], [readOnly],
+  /// [expands], [maxLengthEnforced], [obscureText], [obscureTextBehavior],
+  /// [prefixMode], [readOnly],
   /// [scrollPadding], [suffixMode], [textAlign], [selectionHeightStyle],
   /// [selectionWidthStyle], and [enableSuggestions] properties must not be null.
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_field.dart:399]: Can you adjust the line length of line 397 here?

File: packages/flutter/lib/src/material/material_state_mixin.dart [hunk 19]
```diff
@@ -1,167 +0,0 @@
...
-
-  /// Getter for whether this class considers [MaterialState.selected] to be active.
-  bool get isSelected => materialStates.contains(MaterialState.selected);
-
-  @override
-  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
-    super.debugFillProperties(properties);
-    properties.add(DiagnosticsProperty<Set<MaterialState>>('materialStates', materialStates, defaultValue: <MaterialState>{}));
-  }
-}
```
Comment (justinmc) [inline packages/flutter/lib/src/material/material_state_mixin.dart:167]: Why was this file removed?

File: packages/flutter/lib/src/rendering/editable.dart (hunk 16 above)
Comment (justinmc) [reply packages/flutter/lib/src/rendering/editable.dart:148]: This still looks like a problem here.  There is no obscureTextBehavior, and obscureText is removed.  obscureText should remain and obscureTextBehavior should be added, right?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 20]
```diff
@@ -300,10 +305,12 @@ class CupertinoTextField extends StatefulWidget {
...
        assert(autofocus != null),
        assert(obscuringCharacter != null && obscuringCharacter.length == 1),
-       assert(obscureText != null),
+       assert(obscureText != null || obscureTextBehavior != null),
        assert(autocorrect != null),
-       smartDashesType = smartDashesType ?? (obscureText ? SmartDashesType.disabled : SmartDashesType.enabled),
-       smartQuotesType = smartQuotesType ?? (obscureText ? SmartQuotesType.disabled : SmartQuotesType.enabled),
+       smartDashesType = smartDashesType ?? ((obscureText != null && obscureText) ||
+           (obscureTextBehavior != null && obscureTextBehavior != ObscureTextBehavior.none) ? SmartDashesType.disabled : SmartDashesType.enabled),
+       smartQuotesType = smartQuotesType ?? ((obscureText != null && obscureText) ||
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_field.dart:312]: Instead of checking for the existence of obscureText everywhere, what if we convert it to obscureTextBehavior in the initializers and then only use obscureTextBehavior?

```dart
obscureTextBehavior = obscureTextBehavior ?? (obscureText != null && obscureText) ? ObscureTextBehavior.delayed : ObscureTextBehavior.none,
```

Then I think you can pretty much get rid of obscureText everywhere else.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 21]
```diff
@@ -2951,6 +2951,116 @@ void main() {
...
+        backgroundCursorColor: Colors.grey,
+        obscureTextBehavior: ObscureTextBehavior.delayed,
+        controller: controller,
+        focusNode: focusNode,
+        style: textStyle,
+        cursorColor: cursorColor,
+      ),
+    ));
+
+    final String expectedValue = '•' *controller.text.length;
```
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_test.dart:3025]: This doesn't test the delay, does it?  Maybe if you enter the text via `tester.enterText`, then check that the last character is not `"•"`, then wait a second and test that it has been obscured.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 22]
```diff
@@ -2951,6 +2951,116 @@ void main() {
...
+        backgroundCursorColor: Colors.grey,
+        obscureTextBehavior: ObscureTextBehavior.all,
+        controller: controller,
+        focusNode: focusNode,
+        style: textStyle,
+        cursorColor: cursorColor,
+      ),
+    ));
+
+    final String expectedValue = '•' *controller.text.length;
```
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_test.dart:2970]: Space after the `*` here.

**Review by justinmc — COMMENTED** [review]
Some suggestions for improving this.  There also may be errors from RenderEditable if I'm reading the code right, though I haven't tried running it again.

**Comment by zhangwei911** [comment]
Such a simple function has changed many things for so long, but there are still various mistakes. I decided to give up. Maybe my ability is limited

**Comment by justinmc** [comment]
@zhangwei911 Thanks for the attempt here and for getting us thinking about this.  We should have something like obscureTextBehavior in the future.  Feel free to tag me for review if you open any other PRs.
